### DISCUSSION TITLE
Send ROS2 Commands to Specific IP on Network without Multi-Casting

### DISCUSSION THREAD

**Discussion by paccionesawyer** [iRobotEducation/create3_docs#20] [Questions] [answered]
I'm trying to send ROS2 commands to my Create3 on a network that does not support multicasting. This network has both a "Secure" and "Wireless" version (Something Like a Guest Network). The Create3 is connected to the "Wireless" Network while the computer attempting to send ROS2 commands is connected to the "Secure" Network. 

I have the IP Address, can successfully ping the Create3 and access the webserver it hosts using the "Secure" Network. Is there a way to send ROS2 commands given this setup?

**Comment by shamlian** [comment]
Hello and thank you for the question.
Using ROS 2 in such a setup requires a custom XML profile for your network. At the moment, the robot does not have a facility for uploading such a configuration file, so we'll take this question as a feature request, instead.

**Comment by alsora** [comment, marked as answer]
Hi @paccionesawyer!

We suggest to use Fast-DDS as RMW implementation as it provides documentation for how to deal with particular network configurations such as this one.

What you need to do is the following:
 - make sure that `rmw_fastrtps_cpp` is the selected RMW implementation on the robot by selecting it in the webserver configuration tab (more details here https://iroboteducation.github.io/create3_docs/setup/provision/)
 - make sure that Fast-DDS is installed and it's the selected RMW implementation on your laptop (or Raspberry Pi).
 ```
sudo apt update && sudo apt install ros-galactic-rmw-fastrtps-cpp
export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
 ```
 - Copy this XML configuration file to your laptop (or Raspberry Pi) and replace `ROBOT_IP` with the actual IP of your Create 3 at line 12
 ```
 <?xml version="1.0" encoding="UTF-8" ?>
<profiles xmlns="http://www.eprosima.com/XMLSchemas/fastRTPS_Profiles">
    <participant profile_name="unicast_connection" is_default_profile="true">
        <rtps>
            <builtin>
                <metatrafficUnicastLocatorList>
                    <locator/>
                </metatrafficUnicastLocatorList>
                <initialPeersList>
                    <locator>
                        <udpv4>
                            <address>ROBOT_IP</address>
                        </udpv4>
                    </locator>
                </initialPeersList>
            </builtin>
        </rtps>
    </participant>
</profiles>
```
 - Tell ROS where to find this XML file by doing:
 ```
export FASTRTPS_DEFAULT_PROFILES_FILE=/path/to/the/file
 ```
  - Kill the ROS daemon to make sure that discovery information are refreshed
  ```
  pkill -9 _ros2_daemon
  ```
   - Try to communicate with the robot!
   ```
ros2 topic list
   ```


**NOTE: if Fast-DDS was not already installed on your laptop, you will also have to rebuild the Create 3 ROS 2 messages https://github.com/iRobotEducation/irobot_create_msgs**

**Reply by chrisbuerginrogers** [reply]
Hi all - thanks for the fix - a few questions (1) how do we check to make sure fastrtps is installed properly?  Is it true that if it were not then when you switch the Create, communication would cease?  and (2) can we add that line into the docs for building a rPi image before the building of the msgs? (https://iroboteducation.github.io/create3_docs/setup/pi4/) - I was not clear where that was happening in those docs.

**Reply by shamlian** [reply]
> (2) can we add that line into the docs for building a rPi image before the building of the msgs? 

Already done as of this AM --  see https://github.com/iRobotEducation/create3_docs/pull/21. I also filed an issue here https://github.com/ros2/ros2_documentation/issues/2275 because the ROS 2 Galactic documentation had an error.

> (1) how do we check to make sure fastrtps is installed properly?

As for (1), I'm not sure there is a "good way;" when in doubt, it's easier/faster to just rebuild. Hopefully, now that FastDDS is part of the recommended install in the instructions, people won't run into that problem.

**Reply by chrisbuerginrogers** [reply]
thanks
